Ds4 Speaker Audio Rebase Fix

nickci2002 opened the pull request:
A new PR with a proper rebase.

nickci2002 committed:
Logic update, no QT ui

nickci2002 committed:
Fixing errors

rainmakerv3 committed:
Gui boxes

rainmakerv3 committed:
fixes

rainmakerv3 committed:
prevent device list refreshing too fast when game not running

nickci2002 committed:
Removed duplicate Socket declarations in kernel/file_system.cpp and fs.h

nickci2002 committed:
Fixed clang-format and micDevice errors

nickci2002 committed:
Ran clang-format and fixed rebase compiler issues

nickci2002 committed:
Settings dialog fix

squidbus commented on src/common/config.cpp:
```diff
@@ -194,6 +196,10 @@ static string trophyKey = "";
 
+// Output
+static std::string mainOutputDevice = "Default Device";
+static std::string padSpkOutputDevice = "Default Device";
```
Should use ‘ConfigEntry’ similar to how the mic device does now (and also probably move mic device together with these definitions here). I see you do already have some TODO items for this.

nickci2002 replied:
Done

squidbus commented on src/common/config.cpp:
```diff
@@ -1036,67 +1067,60 @@ void save(const std::filesystem::path& path, bool is_game_specific) {
 
     m_language.setTomlValue(data, "Settings", "consoleLanguage", is_game_specific);
 
-    // All other entries
```
I’m not sure if I’m just looking at the diff wrong but I think this may have been rebased wrong, seems a lot of stuff was changed unrelated.

nickci2002 replied:
This diff is really cursed, but a bit is missing, I'll fix that in my next commit

squidbus commented on src/core/libraries/audio/sdl_audio.cpp:
```diff
@@ -21,8 +21,32 @@ class SDLPortBackend : public PortBackend {
+        } else {
+            try {
+                devId = static_cast<u32>(std::stoul(port_name));
+                if (port_name != SDL_GetAudioDeviceName(devId)) {
```
I don’t see how this can work, how can the config value be both the ID number and device name?

I think what you can do here is loop though 0 to SDL_GetNumAudioDevices(), use SDL_GetAudioDeviceName(index) to get the name of each, and compare that name to the config value until you find a match. That way the config can store the actual device name and keep the right device even if the IDs change.

squidbus reviewed (commented):
I’ll try to keep this moving with reviews so it doesn’t get out of sync like before. Left some comments.

nickci2002 committed:
Addressed squidbus' concerns

nickci2002 committed:
Update config.cpp to adhere to clang-format

nickci2002 committed:
Removed a space causing clang-format to complain

squidbus commented on src/core/libraries/audio/sdl_audio.cpp:
```diff
@@ -21,8 +22,41 @@ class SDLPortBackend : public PortBackend {
+            try {
+                SDL_AudioDeviceID* dev_array = SDL_GetAudioPlaybackDevices(nullptr);
+                for (; dev_array != 0;) {
+                    std::string name = static_cast<std::string>(SDL_GetAudioDeviceName(*dev_array));
```
```suggestion
                    std::string name(SDL_GetAudioDeviceName(*dev_array));
```

Should just construct from the result instead of casting a `const char*` to `std::string`.

squidbus commented on src/core/libraries/audio/sdl_audio.cpp:
```diff
@@ -21,8 +22,41 @@ class SDLPortBackend : public PortBackend {
+                        dev_array++;
+                    }
+                }
+                if (dev_id == 0) {
```
Do we know for sure that the device ID cannot be 0? Maybe should use something like max possible value to denote this if it can be 0.

Should probably also either set stream to nullptr here and return (same as `"None"`) or set `dev_id = SDL_AUDIO_DEVICE_DEFAULT_PLAYBACK;`. I don't really have a preference on the fallback here as long as we do something that makes sense.

nickci2002 replied:
An AudioDeviceID of zero represents an invalid id according to the SDL docs https://wiki.libsdl.org/SDL3/SDL_AudioDeviceID

squidbus reviewed (commented):
Just two more comments from me on the emulation side of things. I'm not as familiar with some of the stuff changing on the Qt side so may need someone else to review that.

nickci2002 committed:
Addressed squidbus' concerns and added fallbacks

Concerns:
- Changed dev_name construct to remove unnecessary cast
- Added an invalid AudioDeviceID macro to replace magic number
